Source/santad/EventProviders/SNTEndpointSecurityManager.mm

```diff
@@ -0,0 +1,304 @@
+/// Copyright 2019 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+
+#import "Source/santad/EventProviders/SNTEndpointSecurityManager.h"
+#include "Source/common/SNTPrefixTree.h"
+
+#import "Source/common/SNTLogging.h"
+
+#include <EndpointSecurity/EndpointSecurity.h>
+#include <bsm/libbsm.h>
+#import <libproc.h>
+
+@interface SNTEndpointSecurityManager ()
+
+@property(nonatomic) es_client_t *client;
+@property(nonatomic) SNTPrefixTree *prefixTree;
+@property (nonatomic, copy) void (^decisionCallback)(santa_message_t);
+@property (nonatomic, copy) void (^logCallback)(santa_message_t);
+
+@end
+
+@implementation SNTEndpointSecurityManager
+
+- (instancetype)init API_AVAILABLE(macos(10.15)) {
+  self = [super init];
+  if (self) {
+    [self establishClient];
+    _prefixTree = new SNTPrefixTree();
+  }
+
+  return self;
+}
+
+- (void)dealloc {
+  delete _prefixTree;
+}
+
+- (void)establishClient API_AVAILABLE(macos(10.15)) {
+  while (!self.client) {
+    es_client_t *client = NULL;
+    es_new_client_result_t ret = es_new_client(&client, ^(es_client_t *c, const es_message_t *m) {
+      [self messageHandler:m];
+    });
+
+    switch (ret) {
+      case ES_NEW_CLIENT_RESULT_SUCCESS:
+        LOGI(@"Connected to EndpointSecurity");
+        self.client = client;
+        return;
+      case ES_NEW_CLIENT_RESULT_ERR_NOT_PERMITTED:
+        LOGE(@"Unable to create EndpointSecurity client, not full-disk access permitted");
+        LOGE(@"Sleeping for 30s before restarting.");
+        sleep(30);
+        exit(ret);
+      default:
+        LOGE(@"Unable to create es client: %d. Sleeping for a minute.", ret);
+        sleep(60);
+        continue;
+    }
+  }
+}
+
+- (void)messageHandler:(const es_message_t *)m API_AVAILABLE(macos(10.15)) {
```

Comment:
Somewhere in here we need to check prefixTree.

Reply:
We do, line 202